internal/webhook/management_webhook.go

```diff
@@ -230,16 +207,67 @@ func getIncompatibleContracts(ctx context.Context, cl client.Client, mgmt *kcmv1
 			return "", fmt.Errorf("not valid ProviderTemplate %s", tplName)
 		}
 
-		for capiVersion := range pTpl.Status.CAPIContracts {
-			if _, ok := capiTpl.Status.CAPIContracts[capiVersion]; !ok {
-				_, _ = incompatibleContracts.WriteString(fmt.Sprintf("core CAPI contract versions does not support %s version in the ProviderTemplate %s, ", capiVersion, pTpl.Name))
+		inUsedProviders, err := getInUsedProviderWithContracts(ctx, cl, pTpl)
+		if err != nil {
+			return "", err
+		}
+
+		exposedContracts := make(map[string]bool)
+		for capiVersion, providerContracts := range pTpl.Status.CAPIContracts {
+			for _, contract := range strings.Split(providerContracts, "_") {
+				exposedContracts[contract] = true
+			}
+			if verifyCapiContract {
+				if _, ok := capiTpl.Status.CAPIContracts[capiVersion]; !ok {
+					_, _ = incompatibleContracts.WriteString(fmt.Sprintf("core CAPI contract versions does not support %s version in the ProviderTemplate %s, ", capiVersion, pTpl.Name))
+				}
+			}
+		}
+
+		if len(inUsedProviders) == 0 {
+			continue
+		}
+		for provider, contracts := range inUsedProviders {
+			for _, contract := range contracts {
+				if !exposedContracts[contract] {
+					_, _ = incompatibleContracts.WriteString(fmt.Sprintf("missing contract version %s for %s provider that is required by one or more ClusterDeployment, ", contract, provider))
+				}
 			}
 		}
 	}
 
 	return strings.TrimSuffix(incompatibleContracts.String(), ", "), nil
 }
 
+func getInUsedProviderWithContracts(ctx context.Context, cl client.Client, pTpl *kcmv1.ProviderTemplate) (map[string][]string, error) {
```

Comment:
nit: should it be like `getInUseProvidersWithContracts`?

Reply:
RIght. Thank you!